### ISSUE TITLE
Should dead code compiler errors be eliminated?

### ISSUE THREAD

**Issue by ben-albrecht** [chapel-lang/chapel#14683]
Should the following programs compile and run?

```chpl
proc f() {
  return true;
  compilerError('this should be dead-code-eliminated');
}
f();
```

```chpl
proc f() throws {
  throw new owned Error();
  compilerError('this should be dead-code-eliminated');
}
f();
```

```chpl
proc f() {
  exit(1);
  compilerError('this should be dead-code-eliminated');
}
f();
```

```chpl
proc f() {
  halt();
  compilerError('this should be dead-code-eliminated');
}
f();
```

Today, all of these compiler errors are emitted rather than dead-code eliminated.

https://github.com/chapel-lang/chapel/issues/14680 exposed a case where a standard library implementation expected the compile errors to be dead-code eliminated.

Is this considered a bug in the compiler, or is this behavior desired?

**Comment by vasslitvinov** [comment]
I consider this an "unimplemented feature".

Removing everything that follows a `return`, a `throw`, or a function with `pragma "function terminates program"` is in the same bucket for me as removing not-taken branches of a param conditional.

**Comment by gbtitus** [comment]
For the `exit()` and `halt()` cases, is there somehow a guarantee that calling either of those functions must put an end to control flow, or is it possible someone could write such code and end up invoking an instance of `exit()` that (say) printed 'Ha ha, I'm not exiting!' and then returned?  If the latter then I don't think we can eliminate the `compilerError()`, because we can't actually tell at compile time whether it's dead.

It's a bit like the following C case, which also mixes compile- and execution-time "control flow", except that I believe the C standard says that things appear as if the preprocessor finishes processing the entire input before the compiler begins doing so.
```
    ...
    exit(1);
#error "Why is this printed?"
    ...
```

**Comment by e-kayrakli** [comment]
Although I don't find the compiler errors triggering in those cases too bothersome:

```chapel
proc foo() {
  return true;
  bar();
}

proc bar() {
  writeln("Hello");
}

writeln(foo());
```

The call `bar()` that comes after `return true` is eliminated (doesn't end up in the C code). Even though `compilerError` is different than that, there is no reason for user to think differently about it. So, it seems to me that the behavior should be consistent and those `compilerError`s (and `compilerWarning`s) should be removed and never be triggered.

For cases other than `return` I think the same high level idea can be applied. As Greg points out there may be complications for those, though.

**Comment by vasslitvinov** [comment]
If a function marked with `pragma "function terminates program"` does not terminate a program, it would be a bug that we should be able to fix.

**Comment by bradcray** [comment]
I agree with Vass here.  I think in eliminating code after an exit() or halt() we wouldn't just be basing it on calling a routine with that name, but that the code is actually calling the exit() or halt() definition that causes program termination.

**Comment by mppf** [comment]
I'd like the above programs to compile and run, personally.

**Comment by mppf** [comment]
I think that #18597 is an example of the same issue.

**Comment by lydia-duncan** [comment]
I think this real world case runs into the same issue:
```chapel
proc channel.seek(region: range(?)) throws {

  if this.locking then
    compilerError("Cannot seek on a locking channel");

  if (!region.hasLowBound()) {
    throw new IllegalArgumentError("region", "must have a lower bound");

  } else { // Without this else, this code complains about the accesses to region.low below, when the if statement above should ensure it is fine
    if (region.hasHighBound()) {
      const err = qio_channel_seek(_channel_internal, region.low, region.high);

      if err then
        throw SystemError.fromSyserr(err);

    } else {
      const err = qio_channel_seek(_channel_internal, region.low, max(int(64)));

      if err then
        throw SystemError.fromSyserr(err);
    }
  }
}

**Comment by vasslitvinov** [comment]
Thanks Lydia.

Also - what strikes me in `proc channel.seek()` above - also in `proc choice()` in Random - is that we throw an IllegalArgumentError when this condition can be detected at compile time. I would like to convert these and other cases like that to calling compilerError().
